infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/wsplugins/BrokersResult.java

```diff
@@ -76,42 +53,34 @@ public void error(Exception e) {
   }
 
   /**
-   * Submits a result of a broker execution.
-   *
-   * <p>It also count down the number of brokers that are waited for the result submission.
+   * Submits the result of a broker execution.
    *
    * @param toolingFromBroker tooling evaluated by a broker that needs to be added into a workspace
-   * @throws InfrastructureException if called more times than {@link #oneMoreBroker()} which
-   *     indicates incorrect usage of the {@link BrokersResult}
+   * @throws InfrastructureException if called second time which indicates incorrect usage of the
+   *     {@link BrokersResult}
    * @throws IllegalStateException if called before the call of {@link #get(long, TimeUnit)}
    */
-  public void addResult(List<ChePlugin> toolingFromBroker) throws InfrastructureException {
+  public void brokerResult(List<ChePlugin> toolingFromBroker) throws InfrastructureException {
     if (!started.get()) {
```

Comment:
Isn't there a race condition here? Consider a thread that runs `PluginBrokerManager#getTooling()` starts the `BrokerEnvironmentFactory#create()`, which I believe then will result in the broker calling the RPC and raising the `BrokerEvent`s (correct me if I am wrong, please) and then gets suspended before it can call `BrokersResult#get()`.

It is possible that broker(s) report(s) back and call this method before the above thread reaches `BrokersResult#get()`.

I know that the argument can be that k8s is going to always be slower than the wsmaster process, but I'm not sure about you but I have seen processes suspended for tens of seconds...

Reply:
`BrokerEnvironmentFactory#create()` creates `KubernetesEnvironment` here https://github.com/garagatyi/che/blob/0f8471014130c72b06c40b2b35137d72f3756dd9/infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/wsplugins/PluginBrokerManager.java#L102 but doesn't deploy anything.
Then we subscribe to events here https://github.com/garagatyi/che/blob/0f8471014130c72b06c40b2b35137d72f3756dd9/infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/wsplugins/PluginBrokerManager.java#L108 (go deeper https://github.com/garagatyi/che/blob/d7f18c7ac3fa09e189b724c4affc8f13637e48ef/infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/wsplugins/brokerphases/ListenBrokerEvents.java#L60)
Then deployment is created here 
https://github.com/garagatyi/che/blob/9dfc0b665411cb9a0d0a55ec5d0a1cdc8705c233/infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/wsplugins/brokerphases/DeployBroker.java#L88 which is lower on stack than https://github.com/garagatyi/che/blob/d7f18c7ac3fa09e189b724c4affc8f13637e48ef/infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/wsplugins/brokerphases/ListenBrokerEvents.java#L62 

Hopefully I haven't missed anything and there is no race condition.

Reply:
Ah, got confused about the difference between create and deploy. Thanks for the explanation.